p2p/src/peer.rs

```diff
@@ -379,38 +388,40 @@ mod run {
                 write,
                 cryptographer,
                 // TODO: eyeball decision of default buffer size of 1 KB, should be benchmarked and optimized
-                buffer: BytesMut::with_capacity(1024),
+                buffer: Vec::with_capacity(1024),
+                queue: BytesMut::with_capacity(1024),
             }
         }
 
-        /// Send byte-encoded message to the peer
+        /// Prepare message for the delivery and put it into the queue to be sent later
         ///
         /// # Errors
         /// - If encryption fail.
-        /// - If write to `stream` fail.
-        async fn send_message<T: Pload>(&mut self, msg: T) -> Result<(), Error> {
+        fn publish_message<T: Pload>(&mut self, msg: T) -> Result<(), Error> {
             // Start with fresh buffer
             self.buffer.clear();
-            let mut writer = (&mut self.buffer).writer();
-            msg.encode_to(&mut writer);
-            let encoded_size = self.buffer.remaining();
-            let encrypted = self.cryptographer.encrypt(&self.buffer[..encoded_size])?;
-            self.buffer.advance(encoded_size);
-            assert!(
-                !self.buffer.has_remaining(),
-                "Buffer must be empty at this point"
-            );
-            let encrypted_size = encrypted.len();
+            msg.encode_to(&mut self.buffer);
+            let encrypted = self.cryptographer.encrypt(&self.buffer)?;
```

Comment:
(not for this PR): Maybe we can get rid of this allocation by encrypting in-place?

Reply:
I was thinking about exactly the same thing, but currently Encryptor trait provide only allocating methods...
Also tried to use only single buffer, but resulting code was tricky and buggy so i‘ve decided to leave it for separate PR.